File: src/features/documents/model/profileSections.ts
import type { ProfileSection } from "@/shared/types/generated/ai";
import type { Profile } from "@/shared/types/generated/profile";

/** Section proposée à l'interrupteur, avec ce que le profil y contient. */
export interface SectionOption {
  readonly section: ProfileSection;
  readonly label: string;
  readonly count: number;
}

/** « Ce que l'IA peut utiliser » pour un CV (`screens/15`). */
export const RESUME_SECTIONS: ReadonlyArray<readonly [ProfileSection, string]> = [
  ["experiences", "Expériences"],
  ["education", "Formations"],
  ["skills", "Compétences"],
  ["languages", "Langues"],
  ["projects", "Projets"],
  ["certifications", "Certifications"],
  ["interests", "Centres d’intérêt"],
];

/**
 * « Arguments autorisés » pour une lettre (`screens/16`) : les faits du profil dont la
 * lettre peut s'appuyer. Les prétentions salariales n'y figurent pas : le profil ne les
 * connaît pas.
 */
export const LETTER_ARGUMENTS: ReadonlyArray<readonly [ProfileSection, string]> = [
  ["summary", "Présentation"],
  ["experiences", "Expériences"],
  ["skills", "Compétences"],
  ["education", "Formations"],
  ["availability", "Disponibilité"],
  ["projects", "Projets"],
  ["certifications", "Certifications"],
];

export function sectionCount(profile: Profile, section: ProfileSection): number {
  switch (section) {
    case "summary":
      return profile.identity.resume?.trim() ? 1 : 0;
    case "availability":
      return profile.identity.availability?.trim() ? 1 : 0;
    case "experiences":
      return profile.experiences.length;
    case "education":
      return profile.education.length;
    case "skills":
      return profile.skills.length;
    case "languages":
      return profile.languages.length;
    case "projects":
      return profile.projects.length;
    case "certifications":
      return profile.certifications.length;
    case "interests":
      return profile.interests.length;
  }
}

export function sectionOptions(
  profile: Profile | null,
  sections: ReadonlyArray<readonly [ProfileSection, string]>,
): SectionOption[] {
  return sections.map(([section, label]) => ({ section, label, count: profile ? sectionCount(profile, section) : 0 }));
}

/** Ajoute ou retire une section de la liste des exclusions. */
export function toggleSection(excluded: readonly ProfileSection[], section: ProfileSection): ProfileSection[] {
  return excluded.includes(section) ? excluded.filter((value) => value !== section) : [...excluded, section];
}

File: src-tauri/src/features/ai/domain/profile_sections.rs
//! Sections du profil que l'utilisateur retire d'une génération (« Ce que l'IA peut
//! utiliser », « Arguments autorisés »).

use crate::features::profile::domain::Profile;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

/// Section du profil. Une section exclue n'est ni envoyée au modèle, ni reprise dans le
/// document, ni proposée ensuite comme contenu à ajouter.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
#[serde(rename_all = "snake_case")]
#[ts(export, export_to = "ai.ts")]
pub enum ProfileSection {
    /// Le résumé de l'identité (« Présentation »).
    Summary,
    /// La disponibilité déclarée dans l'identité.
    Availability,
    Experiences,
    Education,
    Skills,
    Languages,
    Projects,
    Certifications,
    Interests,
}

/// Copie du profil sans les sections exclues. L'identité (nom, coordonnées, titre) reste :
/// elle en-tête le document et ne constitue pas un argument.
#[must_use]
pub fn profile_without(profile: &Profile, excluded: &[ProfileSection]) -> Profile {
    let mut kept = profile.clone();
    for section in excluded {
        match section {
            ProfileSection::Summary => kept.identity.resume = None,
            ProfileSection::Availability => kept.identity.availability = None,
            ProfileSection::Experiences => kept.experiences.clear(),
            ProfileSection::Education => kept.education.clear(),
            ProfileSection::Skills => kept.skills.clear(),
            ProfileSection::Languages => kept.languages.clear(),
            ProfileSection::Projects => kept.projects.clear(),
            ProfileSection::Certifications => kept.certifications.clear(),
            ProfileSection::Interests => kept.interests.clear(),
        }
    }
    kept
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::features::profile::domain::{Education, Experience, Skill};

    fn profile() -> Profile {
        let mut profile = Profile::default();
        profile.identity.first_name = "Jean".into();
        profile.identity.resume = Some("Chargé d'exploitation".into());
        profile.experiences.push(Experience {
            title: "Technicien".into(),
            company: "Ker Informatique".into(),
            ..Experience::default()
        });
        profile.education.push(Education {
            degree: "BTS SIO".into(),
            school: "Lycée Chateaubriand".into(),
            ..Education::default()
        });
        profile.skills.push(Skill {
            name: "Linux".into(),
            ..Skill::default()
        });
        profile
    }

    #[test]
    fn une_section_exclue_disparait_et_les_autres_restent() {
        let kept = profile_without(
            &profile(),
            &[ProfileSection::Skills, ProfileSection::Summary],
        );

        assert!(kept.skills.is_empty());
        assert!(kept.identity.resume.is_none());
        assert_eq!(kept.experiences.len(), 1);
        assert_eq!(kept.education.len(), 1);
        assert_eq!(kept.identity.first_name, "Jean");
    }

    #[test]
    fn sans_exclusion_le_profil_est_intact() {
        assert_eq!(profile_without(&profile(), &[]), profile());
    }
}
